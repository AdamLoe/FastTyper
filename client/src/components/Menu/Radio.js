import React from "react";
import PT from "prop-types";

let RadioItem = ({
	 value, selected,
	 onClick,  mouseEnter, mouseLeave
}) => (
	<button
		value={value}
		onClick={onClick}
		className={"RadioButton " + value + ( selected ? " selected" : "")}
		onMouseEnter={mouseEnter}
		onMouseLeave={mouseLeave}
	>
		{ value }
	</button>
);
RadioItem.propTypes = {
	value: PT.string.isRequired,
	selected: PT.bool.isRequired,
	onClick: PT.func.isRequired,
	mouseEnter: PT.func.isRequired,
	mouseLeave: PT.func.isRequired,
};

class Radio extends React.Component {
	state = {
		hovered: false,
		hoverDescription: ""
	};

	//We wrap mouseEnter with a description prop to not pass desc down
	mouseEnter = (description) =>
		() => this.setState({
			hovered: true,
			hoverDescription: description
		});

	//No need to wrap because we dont need any information
	mouseLeave = () =>
		this.setState({
			hovered: null
		});

	render() {
		let { name, value, updateValue, options } = this.props;
		let { hovered, hoverDescription } = this.state;

		let description = options.map( ([option, description]) => {
			if (option === value) return description;
		});
		if (hovered) description = hoverDescription;

		return (
			<div className="Input InputRadio">
				<div className="InputName">
					{ name }
				</div>
				{ options.map( ([option, description]) => (
					<RadioItem
						key={option} value={option}
						selected={option === value}
						onClick={() => updateValue(option)}
						mouseEnter={this.mouseEnter(description)}
						mouseLeave={this.mouseLeave}
					/>
				))}
				<div className="InputDescription">
					{ description }
				</div>
			</div>
		);
	}

}

Radio.propTypes = {
	name: PT.string.isRequired,
	value: PT.string.isRequired,
	updateValue: PT.func.isRequired,
	options: PT.arrayOf(
		PT.arrayOf(
			PT.string.isRequired
		).isRequired
	).isRequired
};

export default Radio;