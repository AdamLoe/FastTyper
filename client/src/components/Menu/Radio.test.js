import React from "react";
import { create } from "react-test-renderer";
import Radio from "./Radio";

let name = "TestName";
let value = "Default";
let updateValue = val => {
  value = val;
};
let options = [
  ["Default", "10 Per Game"],
  ["Rising", "10 at start. 5 more per game"],
  ["Manual", "Set score each game"]
];

let getComponent = () =>
  create(
    <Radio
      name={name}
      value={value}
      updateValue={updateValue}
      options={options}
    />
  );
let component = getComponent();
let root = component.root;

describe("Radio", () => {
  test("it matches the snapshot", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("updates right values", () => {
    options.map(([option, desc]) => {
      value = null;
      // Name undefined, so we don't include include root
      root.findByProps({ value: option, name: undefined }).props.onClick();
      expect(value).toBe(option);
    });
  });

  test("shows selected description", () => {});
  test("shows hovered description", () => {});

  test("reverts description after unhover", () => {});
  /*
	test("show descriptions", () => {
		options.map(([option, desc]) => {
			root.findByProps({ })
		});
	});
	*/
});
