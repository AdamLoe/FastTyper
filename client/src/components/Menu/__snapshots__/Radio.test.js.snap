// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio it matches the snapshot 1`] = `
<div
  className="Input InputRadio"
>
  <div
    className="InputName"
  >
    TestName
  </div>
  <button
    className="RadioButton Default selected"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    value="Default"
  >
    Default
  </button>
  <button
    className="RadioButton Rising"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    value="Rising"
  >
    Rising
  </button>
  <button
    className="RadioButton Manual"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    value="Manual"
  >
    Manual
  </button>
  <div
    className="InputDescription"
  >
    10 Per Game
  </div>
</div>
`;
